### PR TITLE
Branch python operator decorator

### PR DESCRIPTION
**Pull request by subkanthi**
Branch python decorator very similar to python virtualenv decorator.

```
 @task.branch(task_id="branching")
    def random_choice():
        return random.choice(options)
```

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/decorators/branch_python.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+T = TypeVar("T", bound=Callable)
+
+
+class BranchPythonDecoratorMixin:
+    """
+    Mixin class for adding branch python decorator
+    to the provider managers.
+    """
+
+    def branchpython(
```
Comment (subkanthi) [inline airflow/decorators/branch_python.py]: Maybe branchpython is not a good name.

Comment (uranusjr) [reply airflow/decorators/branch_python.py]: I’d say just call it `branch`.

File: tests/decorators/test_branch_python.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+from airflow.utils.state import State
+
+
+class Test_BranchPythonDecoratedOperator:
+    def test_branch_one(self, dag_maker):
+        @task
+        def dummy_f():
+            pass
+
+        #
```
Comment (uranusjr) [inline tests/decorators/test_branch_python.py]:
```suggestion
```

Accident?

File: tests/decorators/test_branch_python.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            if ti.task_id == "dummy_f":
+                assert ti.state == State.SUCCESS
+            if ti.task_id == "branching":
+                assert ti.state == State.SUCCESS
+            if ti.task_id == "task_1":
+                assert ti.state == State.SUCCESS
+            if ti.task_id == "task_2":
+                assert ti.state == State.SKIPPED
+
+    def test_branch_two(self, dag_maker):
```
Comment (uranusjr) [inline tests/decorators/test_branch_python.py]: These two tests should be merged into one with `@pytest.mark.parametrize`

File: airflow/decorators/branch_python.py (hunk 1 above)
Comment (subkanthi) [reply airflow/decorators/branch_python.py]: Thanks, changed to branch.

File: tests/decorators/test_branch_python.py (hunk 2 above)
Comment (subkanthi) [reply tests/decorators/test_branch_python.py]: Removed, thanks

File: tests/decorators/test_branch_python.py (hunk 3 above)
Comment (subkanthi) [reply tests/decorators/test_branch_python.py]: Thanks merged.

File: airflow/decorators/branch_python.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    Mixin class for adding branch python decorator
+    to the provider managers.
+    """
+
+    def branch(
+        self, python_callable: Optional[Callable] = None, multiple_outputs: Optional[bool] = None, **kwargs
+    ):
+        """
+        Wraps a python function into an Airflow operator to run via a Python virtual environment.
```
Comment (josh-fell) [inline airflow/decorators/branch_python.py]: This should be updated to describe the new decorator.

File: airflow/decorators/branch_python.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        Accepts kwargs for operator kwarg. Can be reused in a single DAG.
+
+        :param python_callable: Function to decorate
+        :type python_callable: Optional[Callable]
+        :param multiple_outputs: if set, function return value will be
+            unrolled to multiple XCom values. List/Tuples will unroll to xcom values
```
Comment (josh-fell) [inline airflow/decorators/branch_python.py]: List/Tuples aren't handled by `multiple_outputs`.

File: airflow/decorators/branch_python.py (hunk 4 above)
Comment (subkanthi) [reply airflow/decorators/branch_python.py]: Fixed.

File: airflow/decorators/branch_python.py (hunk 5 above)
Comment (subkanthi) [reply airflow/decorators/branch_python.py]: Removed, thanks

File: airflow/decorators/branch_python.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+import inspect
+from textwrap import dedent
+from typing import Callable, Optional, Sequence, TypeVar
+
+from airflow.decorators.base import DecoratedOperator, task_decorator_factory
+from airflow.operators.python import BranchPythonOperator
+from airflow.utils.python_virtualenv import remove_task_decorator
```
Comment (subkanthi) [inline airflow/decorators/branch_python.py:24]: This is still referred from python_virtualenv, maybe we can move it a common place for separation of concerns, also used in docker decorator.

Comment (uranusjr) [reply airflow/decorators/branch_python.py:24]: `code_utils` feels like a natural home for the function.

Comment (subkanthi) [reply airflow/decorators/branch_python.py:24]: Thanks @uranusjr , is it OK if I do that in a separate PR since it affects the docker decorator too.

**Review by uranusjr — APPROVED** [review]
This will need some rewrite since we just re-did how a new task decorator should be declared on `_TaskDecorator`. But the `@task.branch` implementation itself is good.

**Comment by subkanthi** [comment]
will wait for https://github.com/apache/airflow/pull/20933 to be merged and will rebase again.

File: airflow/decorators/__init__.pyi [hunk 7]
```diff
@@ -269,6 +287,8 @@ class TaskDecoratorFactory:
     @overload
     def python(self, python_callable: F) -> F: ...
     @overload
+    def branch(self, python_callable: F) -> F: ...
+    @overload
```
Comment (ashb) [inline airflow/decorators/__init__.pyi]: Ah good -- you spotted my change, this will need moving to be next to the other branch.

Comment (ashb) [reply airflow/decorators/__init__.pyi]: @subkanthi My PR is merged now.

Comment (subkanthi) [reply airflow/decorators/__init__.pyi]: Changed, thanks

File: airflow/decorators/__init__.pyi [hunk 8]
```diff
@@ -41,7 +41,6 @@ class TaskDecoratorFactory:
         **kwargs,
     ) -> TaskDecorator:
         """Create a decorator to convert the decorated callable to a task.
-
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: Please don’t remove the blank line (and please add one for `branch`). Those are intended to separate the “summary” and other descriptions.

https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

Comment (subkanthi) [reply airflow/decorators/__init__.pyi]: Restored blank lines.

**Comment by uranusjr** [comment]
Static checks are failing for unrelated errors.

**Comment by subkanthi** [comment]
@potiuk , @uranusjr  can u please check this, the mypy error is not related to the change.

**Comment by potiuk** [comment]
I think it is "triggered" by some of the changes you've done in `cncf.kubrmetes`. MyPy is "stable" (i.e. same files generate same errors) but the stability is a bit "fragile" - i.e. changes in one area might generate new errors discovered "around" - even if they are seemingly unrelate.

In this case it seems that adding new package (the __init_.py in decorators) triggered it. 

I think you do not need that change actually so likely removing it will fix it.

**Comment by subkanthi** [comment]
> I think it is "triggered" by some of the changes you've done in `cncf.kubrmetes`. MyPy is "stable" (i.e. same files generate same errors) but the stability is a bit "fragile" - i.e. changes in one area might generate new errors discovered "around" - even if they are seemingly unrelated.
> 
> In this case it seems that adding new package (the `__init__.py` in decorators) triggered it.
> 
> I think you do not need that change actually so likely removing it will fix it.

Thanks removed it, but looks like it didnt help with the mypy error.

**Comment by potiuk** [comment]
Then I guess you need to fix the error.

**Comment by uranusjr** [comment]
Urgh, I know what’s going on, Mypy doesn’t like it when you reassign a variable to a different type. For this particular case, the erroring file has

https://github.com/apache/airflow/blob/28378d867afaac497529bd2e1d2c878edf66f460/airflow/example_dags/example_kubernetes_executor.py#L61

which assigns a task to a variable types for the task _function_. Mypy somehow allows that (it probably shouldn’t) but errors later when you actually try to use the variable.

**Comment by uranusjr** [comment]
I pushed a commit to your branch that should fix the error.

**Comment by subkanthi** [comment]
Thanks @uranusjr and @potiuk , all green now.

**Review by potiuk — APPROVED** [review]

**Pull request merged**